### ISSUE TITLE
There is insufficient memory for the Java Runtime Environment to continue.

### ISSUE THREAD

**Issue by JawadKM** [balabit/syslog-ng#2298]
Hi ,
I am trying to run syslog-ng but stuck here, would some one be able to assist me in resolving this issue.
Your help will be highly appreciated.

> -- Unit syslog-ng.service has begun starting up.
> Sep 20 17:01:29 ************************* syslog-ng[9542]: OpenJDK 64-Bit Server VM warning: INFO: os::commit_memory(0x00007f9ac80cd000, 2555904, 1) failed; error='Permission denied' (errno=13)
> Sep 20 17:01:29 ************************* syslog-ng[9542]: #
> Sep 20 17:01:29 ************************* syslog-ng[9542]: # There is insufficient memory for the Java Runtime Environment to continue.
> Sep 20 17:01:29 ************************* syslog-ng[9542]: # Native memory allocation (mmap) failed to map 2555904 bytes for committing reserved memory.
> Sep 20 17:01:29 ************************* syslog-ng[9542]: # An error report file with more information is saved as:
> Sep 20 17:01:29 ************************* syslog-ng[9542]: # /tmp/hs_err_pid9542.log
> Sep 20 17:01:29 ************************* systemd[1]: syslog-ng.service: main process exited, code=exited, status=1/FAILURE
> Sep 20 17:01:29 ************************* systemd[1]: Failed to start System Logger Daemon.
> -- Subject: Unit syslog-ng.service has failed

When I further drill down into the issue it gives me this

> [root@opendj syslog-ng]# sudo tail -f /tmp/hs_err_pid9545.log
> 
> Memory: 4k page, physical 8010104k(2211392k free), swap 2097148k(2097148k free)
> 
> vm_info: OpenJDK 64-Bit Server VM (25.181-b13) for linux-amd64 JRE (1.8.0_181-b13), built on Jul 23 2018 15:56:20 by "mockbuild" with gcc 4.8.5 20150623 (Red Hat 4.8.5-28)

looking forward to here from you soon.

**Comment by JawadKM** [comment]
**Version of my syslog-ng**

syslog-ng 3 (3.14.1)
Config version: 3.14
Installer-Version: 3.14.1
Revision:
Compile-Date: Feb 28 2018 07:34:20
Module-Directory: //usr/lib64/syslog-ng
Module-Path: //usr/lib64/syslog-ng
Available-Modules: add-contextual-data,afamqp,affile,afprog,afsocket,afstomp,afuser,appmodel,basicfuncs,cef,confgen,cryptofuncs,csvparser,date,dbparser,disk-buffer,graphite,json-plugin,kvformat,linux-kmsg-format,map-value-pairs,pseudofile,sdjournal,snmptrapd-parser,stardate,syslogformat,system-source,tags-parser,tfgetent,xml,mod-java
Enable-Debug: off
Enable-GProf: off
Enable-Memtrace: off
Enable-IPv6: on
Enable-Spoof-Source: on
Enable-TCP-Wrapper: on
Enable-Linux-Caps: on
Enable-Systemd: on

**Syslog-ng Config for Elasticsearch**

destination d_elastic {
elasticsearch2 (
cluster("syslog-ng")
client-mode("http")
index("syslog-ng")
type("test")
template("$(format-json --scope rfc5424 --scope nv-pairs --exclude DATE --key ISODATE)")
)
};

log {
source(s_network);
destination(d_elastic);
flags(flow-control);
};

**Comment by Kokan** [comment]
@JawadKM were you able to solve this issue ?

**Comment by JawadKM** [comment]
@Kokan I was able to solve the issue, it was a permission issue at SELinux level.

**Comment by Kokan** [comment]
@JawadKM is it possible that the #2292 is also related to selinux ?

**Comment by JawadKM** [comment]
@Kokan yes that issue is also related to SELinux

**Comment by JawadKM** [comment]
@Kokan I stll have a question for you, I will appreciate if you can help me with that.
So for the security purposes I have SELinux **enforcing** and that is how I want it to be, but to run syslog-ng with Elasticsearch2 configuration I had to make SELinux to  **permissive** which I don't want to do in production environment.
Could you please advise how can I make syslog-ng work with elasticsearch2 configuration, without compromising my SELinux mode.
Your help in this regard will be appreciated.

**Comment by Kokan** [comment]
@JawadKM I do not have that much experience with selinux, but I would not advice you to turn off :)
I did some digging myself, first of all created a Fedora VM (which has selinux enforcing by defualt), and compiled syslog-ng there, set up elasticsearch 2.3 in docker, and started syslog-ng with your configuration.

The messages were stored in ES without selinux blocking syslog-ng. My guess would be that your linux distro does not, but Fedore does supply rules for java with selinux by default. I cannot be certain of course as I did not see what error happened, only that the JVM cannot be started with selinux.

Have you tried to start other java application ? Does that work?
What kind of distribution are you using ? Maybe steps to reproduce.

Just a few random links/stuff I thought might be useful for you and bumped into during checking this issue:

* Reference policies: https://github.com/SELinuxProject/refpolicy (it has java)
* audit2why and audit2allow: `cat /var/log/audit/audit.log|audit2why` could help to create your own policies
* https://linux.die.net/man/8/java_selinux
* syslog-ng has se policies: https://github.com/balabit/syslog-ng/tree/master/contrib/selinux (but as far as I could see those do not support java, would be nice to extend after finding the root cause)
